File: packages/lhc/src/inference/classify.ts
// Failure classification (DD-3): a pure table from the boundary's failure
// vocabulary onto Epic 02's retryable-or-not. The queue machinery consumes
// `retryable` exactly as it always has — no queue changes, no new states.
// `safeCall` is the containment wrapper around the host function: a thrown
// exception classifies `other` (AC-3.3) and the adapter-owned timeout race
// classifies `timeout` (DD-6) — both structured failures, so no host
// function behavior can crash a drain.
import type {
  ModelCall,
  ModelCallFailureKind,
  ModelCallInput,
  ModelCallResult,
} from "./types.js";

export const FAILURE_CLASSIFICATION: Record<ModelCallFailureKind, { retryable: boolean }> = {
  rate_limit: { retryable: true },
  timeout: { retryable: true },
  network: { retryable: true },
  empty_output: { retryable: true },
  other: { retryable: true },
  auth: { retryable: false },
  invalid_request: { retryable: false },
};

/** try/catch + timeout race around the host function (AC-3.3, DD-6). */
export async function safeCall(
  call: ModelCall,
  input: ModelCallInput,
  timeoutMs: number,
): Promise<ModelCallResult> {
  let timer: NodeJS.Timeout | undefined;
  const timeout = new Promise<ModelCallResult>((resolve) => {
    timer = setTimeout(() => {
      resolve({
        ok: false,
        kind: "timeout",
        message: `model call timed out after ${String(timeoutMs)}ms`,
      });
    }, timeoutMs);
  });
  // The async wrapper folds synchronous throws into the same rejection path —
  // the host's promise contract is not trusted (AC-3.3).
  const attempt = (async (): Promise<ModelCallResult> => {
    try {
      return await call(input);
    } catch (cause) {
      return { ok: false, kind: "other", message: String(cause) };
    }
  })();
  try {
    // A host that resolves after the race resolves into nothing (DD-6).
    return await Promise.race([attempt, timeout]);
  } finally {
    clearTimeout(timer);
  }
}

File: packages/lhc-py/src/lhc/shared_tech/classify.py
"""Ported from packages/lhc/src/shared-tech/classify.ts.

`safe_call` contains the host function: a thrown exception becomes `other`
and the adapter-owned timeout race becomes `timeout`, so host behavior cannot
crash a drain.
"""

from __future__ import annotations

import asyncio

from .inference_types import ModelCall, ModelCallErr, ModelCallInput, ModelCallResult


# try/catch + timeout race around the host function.
async def safe_call(call: ModelCall, input: ModelCallInput, timeout_ms: int) -> ModelCallResult:
    async def attempt() -> ModelCallResult:
        try:
            # The async wrapper folds synchronous throws into the same path —
            # the host's promise contract is not trusted (may be sync-raising).
            result = call(input)  # type: ignore[operator]
            if asyncio.iscoroutine(result) or asyncio.isfuture(result):
                return await result  # type: ignore[misc]
            return result  # type: ignore[return-value]
        except BaseException as cause:
            # JS `String(cause)` — Error → "Error: msg"; other values → String(value).
            if isinstance(cause, Exception):
                msg = str(cause)
                name = type(cause).__name__
                # V8: String(new Error("x")) === "Error: x"
                if name in ("Error", "Exception") or type(cause) is Exception:
                    message = f"Error: {msg}" if msg else "Error"
                else:
                    message = msg if msg else name
            else:
                message = str(cause)
            return ModelCallErr(kind="other", message=message)

    attempt_task = asyncio.create_task(attempt())
    timeout_task = asyncio.create_task(asyncio.sleep(timeout_ms / 1000.0))
    try:
        done, _pending = await asyncio.wait(
            {attempt_task, timeout_task},
            return_when=asyncio.FIRST_COMPLETED,
        )
        if attempt_task in done:
            return attempt_task.result()
        return ModelCallErr(
            kind="timeout",
            message=f"model call timed out after {timeout_ms}ms",
        )
    finally:
        timeout_task.cancel()
        # Leave attempt running if it lost the race (TS Promise.race behavior);
        # only cancel the timer task.
